src/hotspot/share/gc/serial/cardTableRS.cpp

```diff
@@ -601,20 +344,187 @@ void CardTableRS::verify() {
 CardTableRS::CardTableRS(MemRegion whole_heap) :
   CardTable(whole_heap) { }
 
+// Implemented word-iteration to skip long consecutive clean cards.
+CardTable::CardValue* CardTableRS::find_first_dirty_card(CardValue* const start_card,
+                                                         CardValue* const end_card) {
+  using Word = uintptr_t;
+
+  auto is_word_aligned = [] (const void* const p) {
+      return (((uintptr_t)p) % sizeof(Word)) == 0;
+  };
+
+  static_assert(clean_card_val() == (CardValue)-1, "inv");
+  constexpr Word clean_word = (Word)-1;
+
+  CardValue* i = start_card;
+
+  while (!is_word_aligned(i)) {
+    if (i >= end_card) {
+      return end_card;
+    }
+    if (is_dirty(i)) {
+      return i;
+    }
+    ++i;
+  }
+
+  // Word comparison
+  while (i + sizeof(Word) <= end_card) {
+    Word* i_word = reinterpret_cast<Word*>(i);
+    if (*i_word != clean_word) {
+      // Found sth in this word
+      for (size_t j = 0; j < sizeof(Word); ++j) {
+        if (is_dirty(i + j)) {
+          return i + j;
+        }
```

Comment:
Instead of duplicating the byte search, this could just break out of the word search.